src/main/java/com/example/auctionapp/repository/ProductRepository.java

```diff
@@ -194,6 +202,61 @@ public Double getCollectionLowestPrice(String categoryName) {
         return lowestPrice;
     }
 
+    public Long getProductsCountBySubcategory(Long id) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+
+        CriteriaQuery<Tuple> q = cb.createTupleQuery();
+        Root<Product> resource = q.from(Product.class);
+
+        Predicate predicateForEndDate = getPredicateForEndDate(resource, cb);
+        Predicate predicatedorSubcategory = cb.equal(resource.get("subcategory").get("id"), id);
+
+        q.multiselect(cb.count(resource.get("id")))
+                .where(
+                        cb.and(predicateForEndDate, predicatedorSubcategory)
+                );
+
+        List<Tuple> result = entityManager.createQuery(q).getResultList();
+
+        if(!result.isEmpty()) {
+            return (Long) result.get(0).get(0);
+        }
+
+        return 0L;
+
+    }
+
+    public FilterPriceDto getPricesInfo() {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+
+        CriteriaQuery<Tuple> q = cb.createTupleQuery();
+        Root<Product> resource = q.from(Product.class);
+
+        Predicate predicateForEndDate = getPredicateForEndDate(resource, cb);
+
+        q.multiselect(cb.min(resource.get("price")), cb.max(resource.get("price")), cb.avg(resource.get("price")))
+                .where(predicateForEndDate);
+
+        Tuple result = entityManager.createQuery(q).getResultList().get(0);
+
+        Query pricesQuery = entityManager.createNativeQuery("SELECT count(id), round(cast(price as numeric), -1) FROM public.product GROUP BY price ORDER BY price");
```

Comment:
didnt we say that `function` is viable use case for this

Reply:
Dont just resolve without leaving an explanation why it is resolved.

Reply:
Round function in postgresql looks like round(numeric, integer), and I couldn't find way to convert double to postgres numeric type (any other doesn't work)